// app/stories/modules/user/graphql.js
import pubsub               from '../../apollo/pubsub'

import generateType         from 'engine/graphql/generate/type'
import generateTypeCreate   from 'engine/graphql/generate/typeCreate'
import generateTypeData     from 'engine/graphql/generate/typeData'
import generateTypeUpdate   from 'engine/graphql/generate/typeUpdate'
import generateTypeDelete   from 'engine/graphql/generate/typeDelete'
import generateInputType    from 'engine/graphql/generate/inputType'

import generateMutationList from 'engine/graphql/generate/mutationList'
import generateCustomMutationList from 'engine/graphql/generate/customMutationList'

import generateQueryList    from 'engine/graphql/generate/queryList'
import generateCustomQueryList from 'engine/graphql/generate/customQueryList'

import generateSubscriptionList from 'engine/graphql/generate/subscriptionList'

import generateResolverMap  from 'engine/graphql/generate/resolverMap'

import { add }              from '../../graphql'

import config               from './config'
import model                from './model'

export const type = generateType(config)
export const typeData = generateTypeData(config)
export const typeCreate = generateTypeCreate(config)
export const typeUpdate = generateTypeDelete(config)
export const typeDelete = generateTypeUpdate(config)
export const inputType = generateInputType(config)

export const mutationList = generateMutationList(config)
export const customMutationList = generateCustomMutationList(config)

export const queryList = generateQueryList(config)
export const customQueryList = generateCustomQueryList(config)

export const subscriptionList = generateSubscriptionList(config)

import customResolverMap    from './resolvers'

export const resolverMap  = generateResolverMap({
  ...config,
  model,
  pubsub,
  customResolverMap
})

add({
  mutationList: [ ...mutationList, ...customMutationList ],
  resolverMap,
  subscriptionList,
  typeList: [
    type,
    typeData,
    typeCreate,
    typeUpdate,
    typeDelete,
    inputType,
  ],
  queryList: [ ...queryList, ...customQueryList ],
})
